pkg/s3select/select.go

```diff
@@ -276,91 +297,111 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
 					}
 				}
 			}
+
 			filtrCount++
 		}
 		counter++
 	}
+
+}
+
+//ReadIndividualRecord return a map[string]interfaces{} for each row in csv and each json for JSON document
+func readIndividualRecord(inputType SelectQuery) (map[string]interface{}, error) {
+	var individualRecord map[string]interface{}
+	var err error
+	switch inputType.(type) {
+	case *CSVInput:
+		individualRecord, err = inputType.Read()
+	case *JSONInput:
+		individualRecord, err = inputType.Read()
+	}
+	// In case an exception arises while parsing
+	if err != nil {
+		return nil, err
+	}
+	return individualRecord, nil
 }
 
 // printAsterix helps to print out the entire row if an asterix is used.
-func (reader *Input) printAsterix(record []string) string {
+func printAsterix(record []string, inputType SelectQuery) string {
 	myRow := record[0]
 	for i := 1; i < len(record); i++ {
-		myRow = myRow + reader.options.OutputFieldDelimiter + record[i]
+		myRow = myRow + inputType.getOutPutFieldDelimitter() + record[i]
 	}
 	return myRow
 }
 
 // processColumnNames is a function which allows for cleaning of column names.
-func (reader *Input) processColumnNames(reqColNames []string, alias string) error {
-	for i := 0; i < len(reqColNames); i++ {
-		// The code below basically cleans the column name of its alias and other
-		// syntax, so that we can extract its pure name.
-		reqColNames[i] = cleanCol(reqColNames[i], alias)
+func processColumnNames(reqColNames []string, alias string, inputType SelectQuery) error {
+	switch inputType.(type) {
+	case *CSVInput:
+		for i := 0; i < len(reqColNames); i++ {
+			// The code below basically cleans the column name of its alias and other
+			// syntax, so that we can extract its pure name.
+			reqColNames[i] = cleanCol(reqColNames[i], alias)
+		}
+
+	case *JSONInput:
+		// JSON doesnt have columns so no cleaning
 	}
+
 	return nil
 }
 
 // processColNameIndex is the function which creates the row for an index based
 // query.
-func (reader *Input) processColNameIndex(record []string, reqColNames []string, columns []string) (string, error) {
```

Comment:
Why is the record changed from a slice to a string? Was it wrong before?

Reply:
We dont get the record as slice now because the code is made generics, We have map[string]interface{} which is decoded to string and thus the record is a string now.